library/core/src/pin.rs

```diff
@@ -1,188 +1,621 @@
+//! [`Copy`]ing a value from one place in memory to another. In Rust, "move" carries with it the
+//! semantics of ownership transfer from one variable to another, which is the key difference
+//! between a [`Copy`] and a move. For the purposes of this module's documentation, however, when
+//! we write *move* in italics, we mean *specifically* that the value has *moved* in the mechanical
+//! sense of being located at a new place in memory.
+//!
+//! All values in Rust are trivially *moveable*. This means that the address at which a value is
+//! located is not necessarily stable in between borrows. The compiler is allowed to *move* a value
+//! to a new address without running any code to notify that value that its address
+//! has changed. Although the compiler will not insert memory *moves* where no semantic move has
+//! occurred, there are many places where a value *may* be moved. For example, when doing
+//! assignment or passing a value into a function.
+//!
+//! ```
+//! #[derive(Default)]
+//! struct AddrTracker(Option<usize>);
+//!
+//! impl AddrTracker {
+//!     // If we haven't checked the addr of self yet, store the current
+//!     // address. If we have, confirm that the current address is the same
+//!     // as it was last time, or else panic.
+//!     fn check_for_move(&mut self) {
+//!         let current_addr = self as *mut Self as usize;
+//!         match self.0 {
+//!             None => self.0 = Some(current_addr),
+//!             Some(prev_addr) => assert_eq!(prev_addr, current_addr),
+//!         }
+//!     }
+//! }
+//!
+//! // Create a tracker and store the initial address
+//! let mut tracker = AddrTracker::default();
+//! tracker.check_for_move();
+//!
+//! // Here we shadow the variable. This carries a semantic move, and may therefore also
+//! // come with a mechanical memory *move*
+//! let mut tracker = tracker;
+//!
+//! // May panic!
+//! // tracker.check_for_move();
+//! ```
+//!
+//! In this sense, Rust does not guarantee that `check_for_move()` will never panic, because the
+//! compiler is permitted to *move* `tracker` in many situations.
+//!
+//! Common smart-pointer types such as [`Box<T>`] and [`&mut T`] also allow *moving* the underlying
+//! *value* they point at: you can move out of a [`Box<T>`], or you can use [`mem::replace`] to
+//! move a `T` out of a [`&mut T`]. Therefore, putting a value (such as `tracker` above) behind a
+//! pointer isn't enough on its own to ensure that its address does not change.
+//!
+//! # What is "pinning"?
+//! [what-is-pinning]: self#what-is-pinning
+//!
+//! We say that a value has been *pinned* when it has been put into a state where it is guaranteed
+//! to remain *valid* and *located at the same place in memory* from the time it is pinned until its
+//! [`drop`] is called.
+//!
+//! ## Address-sensitive values, AKA "when we need pinning"
+//! [address-sensitive-values]: self#address-sensitive-values-aka-when-we-need-pinning
```

Comment:
I really like this section. However, it's also basically re-telling the entire story of what led to `Pin` and pointing out all the alternatives that don't work for us. Is that really the best way to explain `Pin`? I think ideally we'd be able to explain the way things are without detouring through non-working alternatives, and then offer a supplementary section at the end for those that are curious that discusses a bit the possible alternatives.

Reply:
Yeah I do think we may be able to move this to the bottom. I think we should try doing this in a followup, though.

Reply:
I'm just worried having this at the top of the documentation will cause needless distraction, people thinking they have to understand this to be able to use `Pin`.

Reply:
The above text ("users of Pin in safe code do not need to read it in detail") already suggests that these sections may be skipped/skimmed if you are user of Pin, so I don't think that it's necessarily a distraction.   I also don't think it's position is all that consequential, and so I think it could easily be moved downward.